cmd/main_test.go

```diff
@@ -0,0 +1,145 @@
+// Copyright 2024 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"log/slog"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/thejerf/slogassert"
+
+	"github.com/abcxyz/abc-updater/pkg/metrics"
+	"github.com/abcxyz/abc-updater/pkg/server"
+	"github.com/abcxyz/pkg/logging"
+	"github.com/abcxyz/pkg/renderer"
+)
+
+// Assert testMetricsDB satisfies pkg.MetricsLookuper.
+var _ server.MetricsLookuper = (*testMetricsDB)(nil)
+
+type testMetricsDB struct {
+	apps map[string]*server.AppMetrics
+}
+
+// Update is a Noop.
+func (db *testMetricsDB) Update(ctx context.Context, params *server.MetricsLoadParams) error {
+	return nil
+}
+
+func (db *testMetricsDB) GetAllowedMetrics(appID string) (*server.AppMetrics, error) {
+	if db.apps == nil {
+		// TODO: this should probably log an error and bubble up as a 5xx
+		return nil, fmt.Errorf("no metric definition found for app %s", appID)
+	}
+	v, ok := db.apps[appID]
+	// TODO: this should bubble up as a 404
+	if !ok {
+		return nil, fmt.Errorf("no metric definition found for app %s", appID)
+	}
+	return v, nil
+}
+
+func marshalRequest(tb testing.TB, req *metrics.SendMetricRequest) io.Reader {
+	tb.Helper()
+	b, err := json.Marshal(req)
+	if err != nil {
+		tb.Fatalf("could not marshal json: %s", err.Error())
+	}
+	return bytes.NewReader(b)
+}
+
+func Test_handleMetric(t *testing.T) {
+	t.Parallel()
+	cases := []struct {
+		name       string
+		db         server.MetricsLookuper
+		body       io.Reader
+		wantStatus int
+		wantLogs   map[*slogassert.LogMessageMatch]int
+	}{
+		// TODO: more test cases if by some miracle this ugly thing doesn't get ousted in code review
+		{
+			name: "happy_single_metric",
+			db: &testMetricsDB{apps: map[string]*server.AppMetrics{"test": {
```

Comment:
Typically you'd break this up over multiple lines. 3 open braces on the same line is perhaps a bit much

Reply:
Broken up. If only golang had sets that would have a super clean oneliner